### PR TITLE
fix(tempo): coordinate parallel sponsor fee exposure

### PR DESCRIPTION
**Pull request by gakonst**
## What changed

- replace sender-level serialization with a sponsor-wide atomic fee budget
- key the budget by the recovered on-chain sponsor address, independent of replay-store prefixes
- persist the final sponsor-signed transaction hash and an owner-fenced lifecycle (`prepared` → `broadcasting` → `pending`)
- keep broadcast/pending exposure reserved until a receipt is observed or the expiring nonce reaches `validBefore`
- reconcile receipts lazily before admitting more work; stale workers cannot release replacement reservations
- bound both aggregate declared fee exposure and reservation count
- wait for capacity until the payment challenge deadline rather than rejecting parallel expiring-nonce submissions

## Why

Tempo expiring nonces are designed for independent transactions to be submitted in parallel. A sender mutex defeats that property, while releasing budget immediately after an optimistic RPC response fails to bound transactions that are still pending or whose response was lost.

This keeps the aggregate sponsor bound without coupling independent senders or same-sender expiring-nonce transactions. The final signed hash and owner token provide the useful lifecycle/reconciliation pieces discussed in #690; this PR intentionally does **not** take #690’s sender lock or add a manual recovery API. It supersedes #690 for this behavior.

Tempo references:
- https://docs.tempo.xyz/guide/payments/send-parallel-transactions
- https://docs.tempo.xyz/protocol/transactions

## Policy

`tempo.charge({ feePayerPolicy })` accepts:

- `maxInFlightTotalFee`: aggregate declared worst-case fee exposure; defaults to `maxTotalFee * 10`
- `maxInFlightReservations`: count bound protecting the budget from zero/low-fee request amplification; defaults to `100`

Per-transaction fee validation remains in the shared fee-payer policy. The aggregate settings are charge-broadcast policy and do not leak into session fee-payer configuration. A single transaction larger than the aggregate budget is rejected immediately instead of waiting until expiry.

Capacity waiters perform atomic no-op reads while full. A challenge expiry stops a waiter from beginning new work, but an already broadcast reservation remains counted through the transaction’s later `validBefore` boundary.

## Validation

- reproduced the previous same-sender rejection on pristine `main`
- verified concurrent same-sender expiring-nonce transactions succeed
- verified different senders sharing one sponsor coordinate against the same atomic budget
- verified capacity waits and proceeds after terminal receipt reconciliation
- verified optimistic sponsored broadcasts remain persisted as `pending`
- verified owner fencing rejects stale transitions/releases
- verified reservation-count limiting and oversized-transaction rejection
- local Tempo node: 180 focused fee-payer/charge/budget tests passed
- `pnpm check:ci`
- `pnpm exec tsgo -b`
- `pnpm build`

Both commits are SSH-signed for GitHub verification.

### CODE REVIEW

File: src/tempo/server/Charge.ts [hunk 1]
```diff
@@ -642,8 +659,8 @@ export function charge<const parameters extends charge.Parameters>(
             if (releaseReservation) await releaseHashUse(store, hash)
             throw error
           } finally {
-            if (sponsoredSenderReservation)
-              await releaseSponsoredSenderInFlight(store, sponsoredSenderReservation)
+            if (sponsorBudgetReservation)
+              await releaseSponsorBudget(store, sponsorBudgetReservation)
```
Comment (parvahuja) [inline src/tempo/server/Charge.ts]: with `waitForConfirmation: false`, this releases the budget while the transaction is still pending, so the sponsor could exceed the limit - maybe we should hold it until the transaction confirms or expires, although that may be slightly overkill

**Comment by parvahuja** [comment]
outside of the edge case above, i think 👍

File: src/tempo/server/Charge.test.ts [hunk 2]
```diff
@@ -1939,7 +1950,7 @@ describe('tempo', () => {
       httpServer.close()
     })
 
-    test('behavior: fee payer rejects concurrent in-flight transactions from one sender', async () => {
+    test('behavior: fee payer allows concurrent expiring-nonce transactions from one sender', async () => {
```
Comment (brendanjryan) [inline src/tempo/server/Charge.test.ts:1953]: 👍

File: src/tempo/server/Charge.ts [hunk 3]
```diff
@@ -1095,12 +1117,29 @@ function getProofStoreKey(challengeId: string): `mppx:charge:${string}` {
...
+type SponsorBudgetState = {
+  reservations: Record<string, { expiresAt: number; fee: string }>
+}
+
+type SponsorBudgetReservation = {
+  chainId: number
+  expiresAt: number
+  fee: bigint
+  hash: `0x${string}`
+  sponsor: string
```
Comment (brendanjryan) [inline src/tempo/server/Charge.ts]: is this an address as well? I think we could do the same `0x$` typing?

File: src/tempo/server/Charge.ts [hunk 4]
```diff
@@ -1136,23 +1175,85 @@ function parseHashCredentialSource(parameters: {
...
-): Promise<boolean> {
-  return store.update(getSponsoredSenderStoreKey(parameters), (current) => {
-    if (current !== null) return { op: 'noop', result: false }
-    return { op: 'set', value: Date.now(), result: true }
-  })
+  parameters: SponsorBudgetReservation & { maxInFlightTotalFee: bigint },
+): Promise<void> {
+  const key = getSponsorBudgetStoreKey(parameters)
+  let pollIntervalMs = sponsorBudgetPollIntervalMs
+  for (;;) {
```
Comment (brendanjryan) [inline src/tempo/server/Charge.ts]: 😬 can we add a comment that this won't infinitely spin because of :1196? as long as a reasonably expiresAt is set?

File: src/tempo/server/Charge.ts [hunk 5]
```diff
@@ -1136,23 +1175,85 @@ function parseHashCredentialSource(parameters: {
...
+    const now = Date.now()
+    if (now >= parameters.expiresAt)
+      throw new VerificationFailedError({
```
Comment (brendanjryan) [inline src/tempo/server/Charge.ts]: would feel better if this had more tests

File: src/tempo/server/Charge.ts (hunk 1 above)
Comment (gakonst) [reply src/tempo/server/Charge.ts]: Fixed in a034bbef. Optimistic broadcasts now persist the final sponsor-signed transaction hash and remain budgeted in the `pending` phase until receipt reconciliation or the transaction reaches `validBefore`; ambiguous broadcast failures remain in `broadcasting` under the same rule. The new localnet test asserts the optimistic reservation is still present after the RPC response.

File: src/tempo/server/Charge.ts (hunk 3 above)
Comment (gakonst) [reply src/tempo/server/Charge.ts]: Yep. The hosted fee-payer response is now cryptographically recovered to its concrete `Hex` sponsor address, and local sponsorship already supplies the account address. Budget keys are therefore always scoped by `chainId + recovered sponsor address`; the URL/hash fallback was removed.

File: src/tempo/server/Charge.ts (hunk 4 above)
Comment (gakonst) [reply src/tempo/server/Charge.ts]: Fixed and documented in the helper. Admission waits are bounded by `waitUntil = min(challenge expiry, validBefore)`, so the loop terminates instead of spinning indefinitely. Once broadcast, the separate persisted `expiresAt = validBefore` keeps exposure counted for the full period in which the transaction can still land.

File: src/tempo/server/Charge.ts (hunk 5 above)
Comment (gakonst) [reply src/tempo/server/Charge.ts]: Added focused tests for pending-until-receipt reconciliation, stale-owner fencing, reservation-count limiting, and immediate rejection when one transaction exceeds the aggregate budget. Charge integration tests also cover same-sender parallelism, shared sponsor capacity waiting, expired cleanup, recovered sponsor scoping, and optimistic pending persistence. The focused local Tempo suite passes.
